qualtran/symbolics/math_funcs.py

```diff
@@ -82,8 +83,8 @@ def smin(*args):
     return min(*args)
 
 
-def prod(*args: SymbolicInt) -> SymbolicInt:
-    ret: SymbolicInt = 1
+def prod(args: Iterable[AnySymbolic]) -> AnySymbolic:
+    ret: AnySymbolic = 1
```

Comment:
I think it's better to define an @overload of this method so that if all inputs are `Iterable[SymbolicInt]` then the return type should be a `SymbolicInt` and if the input is `Iterable[SymbolicFloat]` then the return type should be `SymbolicFloat`. 

Using `AnySymbolic` would mean that call sites that expect a `SymbolicInt` would now need to deal with `Union[SymbolicFloat, SymbolicInt]`  instead. I don't think `AnySymbolic` is a useful addition since `SymbolicFloat` is a superset of `SymbolicInt` so the union might as well just correspond to `SymbolicFloat`.

Reply:
@charlesyuan314 wrote 

> here does not seem to be a correct approach using @overload instead due to limitations of mypy, which does not consider float | Expr and int | Expr to be two valid overloads for a function since int is automatically promoted to float.

did you see this?

Reply:
This is where I got stuck @tanujkhattar, if you have a suggestion for how to make `@overload` work:

https://mypy-play.net/?mypy=latest&python=3.12&flags=show-error-codes%2Callow-redefinition%2Ccheck-untyped-defs

I tried a couple of other things, but mypy is pretty stubborn with what it considers a valid overload.

Reply:
> Using AnySymbolic would mean that call sites that expect a SymbolicInt would now need to deal with Union[SymbolicFloat, SymbolicInt]

To my knowledge, this isn't true. The type checker automatically instantiates `AnySymbolic` to be `SymbolicInt` when `prod` is called with a `SymbolicInt`. One way to think about the situation is that there are two versions of `prod` (call them `prod_int` and `prod_float`), each one of which is selected automatically behind the scenes without the caller needing to care about it.

> I don't think AnySymbolic is a useful addition since SymbolicFloat is a superset of SymbolicInt so the union might as well just correspond to SymbolicFloat.

`AnySymbolic` is not really intended to be used by itself as a replacement for the other symbolic types. It's just a type checking tool used in the type signature of `prod` (and possibly similar functions) to get around this specific behavior of the type checker. I'm very open to a way of making the typing work without this `TypeVar` and with `@overload` alone, but I haven't found one yet.

Reply:
Perhaps we can rename AnySymbolic to `SymbolicT` which is the convention for the other TypeVars used as generic types elsewhere in the library (and in Python more generally)

Reply:
Renamed.